src/http2.ts

```diff
@@ -0,0 +1,224 @@
+ * @param config - Request options.
+ */
+export async function request<T>(
+  config: GaxiosOptions
+): Promise<GaxiosResponse<T>> {
+  // Make sure users know this API is unstable
+  if (!warned) {
+    const message = `
+      The HTTP/2 API in googleapis is unstable! This is an early implementation
+      that should not be used in production.  It may change in unpredictable
+      ways. Please only use this for experimentation.
+    `;
+    process.emitWarning(message, 'GOOG_HTTP2');
+    warned = true;
+  }
+
+  const opts = extend(true, {}, config);
+  opts.validateStatus = opts.validateStatus || validateStatus;
+  opts.responseType = opts.responseType || 'json';
+
+  const url = new URL(opts.url!);
+
+  // Check for an existing session to this host, or go create a new one.
+  const session = _getClient(url.host);
+
+  // Since we're using this session, clear the timeout handle to ensure
+  // it stays in memory and connected for a while further.
+  if (session.timeoutHandle !== undefined) {
+    clearTimeout(session.timeoutHandle);
+  }
+
+  // Assemble the querystring based on config.params.  We're using the
+  // `qs` module to make life a little easier.
+  let pathWithQs = url.pathname;
+  if (config.params) {
+    const q = qs.stringify(opts.params);
+    pathWithQs += `?${q}`;
+  }
+
+  // Assemble the headers based on basic HTTP2 primitives (path, method) and
+  // custom headers sent from the consumer.  Note: I am using `Object.assign`
+  // here making the assumption these objects are not deep.  If it turns out
+  // they are, we may need to use the `extend` npm module for deep cloning.
+  const headers = Object.assign({}, opts.headers, {
+    [HTTP2_HEADER_PATH]: pathWithQs,
+    [HTTP2_HEADER_METHOD]: config.method || 'GET',
+    [HTTP2_HEADER_CONTENT_TYPE]: 'application/json',
+  });
+
+  const req = session.client.request(headers);
+  const chunks: Buffer[] = [];
+  const res: GaxiosResponse<T> = {
+    config,
+    request: (req as {}) as GaxiosXMLHttpRequest,
+    headers: [],
+    status: 0,
+    data: {} as T,
+    statusText: '',
+  };
+
+  return new Promise((resolve, reject) => {
+    req
+      .on('response', headers => {
+        res.headers = headers;
+        res.status = Number(headers[HTTP2_HEADER_STATUS]);
+        let stream: Readable = req;
+        if (headers[HTTP2_HEADER_CONTENT_ENCODING] === 'gzip') {
+          stream = req.pipe(zlib.createGunzip());
+        }
+        if (opts.responseType === 'stream') {
+          res.data = (stream as {}) as T;
+          resolve(res);
+          return;
+        }
+        stream
+          .on('data', d => {
+            chunks.push(d);
+          })
+          .on('error', err => {
+            reject(err);
+            return;
+          })
+          .on('end', () => {
+            let data = Buffer.concat(chunks).toString('utf8');
+            try {
+              data = JSON.parse(data);
+            } catch {
+              // assume we got text back
```

Comment:
does the content type header help here, to avoid the guessing game?

Reply:
I made this more consistent with the somewhat sketchy implementation in `gaxios` which also doesn't sniff the content type header.